### DISCUSSION TITLE
Resurrection of the target-, and maybe, host platform S/370 (MVS&CMS)

### DISCUSSION THREAD

**Discussion by mgrossmann** [open-watcom/open-watcom-v2#651] [Q&A]
Hello everyone,

the Watcom C compiler was originally developed for the mainframe, the S/370 platform. Supported operating systems where MVS and CMS/VM.  Unfortunately, Watcom has not developed this platform further over time. 

But since I have seen that parts of the implementation are still available (see [bld/cg/s37](https://github.com/open-watcom/open-watcom-v2/tree/master/bld/cg/s37)), my question to you would be, how big do you estimate the effort to bring this platform back to life. Which steps would you have to take to reach this goal?

Many thanks and best wishes
Mike

**Comment by jmalak** [comment]
I don't know anything about MVS and CMS/VM platforms.
I think that support for S/370 systems was based on some IBM specific processor.
Now-days this processor (32-bit) doesn't exists and is not used anymore.
Old S/370 processor support is absolutely unmaintain.
Also stuff for MVS and CMS/VM platforms as C run-time library and file system don't exist.

I think that effort for it is as implement new architecture and new OS from begining.

**Comment by mgrossmann** [comment]
That is absolutely right. The S/370 processor architecture is not used today. 
Correctly one would have to say, today the successor architecture is used, which is downward compatible however. 

Unfortunately I don't know the structure of this compiler, but what exactly does the part below cg/ do? 
It seems to be a code generator. How or where does this play a role in the compiler? 

How would you estimate the effort to implement a completely new operating system or platform? ( except the rt library)
Where can I find other implementations of rt libs in the source tree? It is bld/clib, right?

**Comment by jmalak** [comment]
We hold this code in bld/cg/s37/... for idea how can be implemented some specific architecture.

Anyway, all OW compilers are divided to several part.
Core common part is Code Generator (bld/cg/... subdirectory). It has common code in bld/cg/c and bld/cg/h and architecture specific code in bld/cg/...arch../...
Next part is code for each type of compiler is Front End which use common Code Generator.
There are C front end in bld/c/...
C++ front end in bld/plusplus/...
Fortran front end in bld/f77/wfc/...
For creation of executable you need run-time-libraries
C run-time library in bld/clib/...
C++ run-time library in bld/cpplib/...
Fortran run-time library in bld/f77/f77lib/...
Math run-time library in bld/mathlib/...

I have no experience with implementation of new architecture that I am not able to estimate.

**Comment by jmalak** [comment]
OW compilers are cross-compilers that you can compile on any supported host platform for any supported target platform.
host platforms are different then target platforms. Generally target platforms count is higher then host platforms.

**Comment by mgrossmann** [comment]
@jmalak Thank you for your explanations about the compiler structure.

so for a test drive, I will have to change the c compiler frontend to use the s37 codegen, right?

/mike

**Comment by mgrossmann** [comment]
Do you still have contacts with Sybase who provided you with the code back then? Or can you give me a contact person there?

**Comment by jmalak** [comment]
If you want recover s37 code compilers then you need to recover Code Generator as first to generate s37 architecture code.
As soon as Code Generator for s37 will be working then you create new architecture configuration to compile compilers with appropriate Code Generator. You don't need to compile on S/370 system you can create cross-compiler by example for Windows or Linux which will generate code for s37 architecture.
Front end should not be dependent on architecture only a few minor parts as #pragma etc. the main part is Code Generator.

**Comment by jmalak** [comment]
No I am not in contact with SyBase any way. Now SyBase is owned by SAP.

**Comment by mgrossmann** [comment]
I am currently trying to build the whole project on my MAC. If that works, what do I have to tell the owcc to use the s37 code generator? march=370? I'm using the time of the build process to get to the bottom of this question.

Thanks for your support

**Comment by jmalak** [comment]
owcc tool is only driver for OW tools as compiler, assembler, librarian, linker, disassembler etc.
If you get compiler working then you can add appropriate option to owcc.
By example C compiler name is depend on target architecture
16-bit Intel            wcc
32-bit Intel            wcc386
alpha                     wccaxp
Power PC               wccppc
....
that appropriate name should be
S/370                     wccs37
that owcc option should be
-march=s37

**Comment by mgrossmann** [comment]
ty

**Comment by mgrossmann** [comment]
How can I disable building unnecessary stuff?

> rc wresui.res
cl setup.exe
bind wresui setup.exe
============ 16:47:59 /Users/mike/work/open-watcom-v2/bld/setupgui ============
**** BUILD rule
============ 16:47:59 /Users/mike/work/open-watcom-v2/bld/docstxt =============
========== 16:47:59 /Users/mike/work/open-watcom-v2/bld/docstxt/docs ==========
wsplice -kC ../readme.sp readme.txt
wsplice -kF77 ../readme.sp freadme.txt
wsplice ../areadme.sp areadme.txt
============ 16:47:59 /Users/mike/work/open-watcom-v2/bld/docstxt =============
============ 16:47:59 /Users/mike/work/open-watcom-v2/bld/docstxt =============
**** BUILD rule
============== 16:47:59 /Users/mike/work/open-watcom-v2/bld/wgml ==============
======= 16:47:59 /Users/mike/work/open-watcom-v2/bld/wgml/test/testlib ========
/Users/mike/work/open-watcom-v2/build/mif/wgmlcmd.mif(75): Error(E33): Unsupported building platform !!!!!!
Error(E02): Make execution terminated
<pmake -d build guitool .not .and         -h> => non-zero return: 512
Build failed

setvars.sh
`export OWDOCBUILD=0                                                     
 export OWGUINOBUILD=1                                                   
 export OWNOBUILD=ide                                                    
 export OWDISTRBUILD=0  `

**Comment by jmalak** [comment]
You need to install DOSBOX because we have one executable for DOS without source code.

NOTE: OW is not ported to OSX (C run-time library).
Only OW build system is ported to OSX (use CLang compiler for bootstrap).
You are able to build OW on OSX, but you can not build executable for OSX by OW.

**Reply by mcejp** [reply]
From https://github.com/open-watcom/open-watcom-v2/issues/251#issuecomment-224388275 it would seem that not so long ago, it was indeed possible to build OW without Dosbox/dosemu by omitting docs and GUI components.
Is this not the case anymore?

**Reply by jmalak** [reply]
It should be working now with setup
OWDOCBUILD=0
OWGUINOBUILD=1

**Reply by mcejp** [reply]
For future reference, @jmalak 's commit is 6eda89da37e34e0c3b00b143189d9b4c81650869
